### PR TITLE
timer: cortex_m: Fix the discontinuous readings of the cycle counter.

### LINKED ISSUE #6164: timer: cortex_m: Incorrect read of clock cycles counter after idle tickless period
On a stm32, I found that when we uses the k_busy_wait function to wait a delay (500us) just after a idle tickless period the delay could be lower than the requested one. I found that consecutive readings of the cycle counter made with the function k_cycle_get_32 juste after tickeless period provided erroneous values (value jump) after some time (100 us).

### PR DESCRIPTION
**Pull request by ghost**
On a stm32, When we use the k_busy_wait function to wait a
delay (500us) just after an idle tickless period the delay could be
lower than the requested one. Consecutive readings of the cycle counter
made with the function k_cycle_get_32 juste after tickeless period
provides erroneous values (value jump) after some time (100 us).

To fix this issue we needs:

- Add the update of clock_accumulated_count value in the
_timer_idle_exit function.

- Treat the case in the get_elapsed_count function when the reload value
of the timer is set to a remaining value to wait until end of tick (see
_timer_idle_exit) . In this case the time elapsed until the systick
timer restart was not yet added to clock_accumulated_count. To retrieve
a correct cycle count we must therefore consider the number of cycle
since current tick period start and not only the cycle number since the
timer restart.

Fixes #6164

Signed-off-by: Holman Greenhand <greenhandholman@gmail.com>

### CODE REVIEW

File: drivers/timer/cortex_m_systick.c [hunk 1]
```diff
@@ -753,7 +755,25 @@ return (u32_t) get_elapsed_count();
 
 	do {
 		cac = clock_accumulated_count;
+#ifdef CONFIG_TICKLESS_IDLE
+		/* When we leave a tickless period the reload value of the timer
+		 * can be set to a remaining value to wait until end of tick.
+		 * (see _timer_idle_exit). The remaining value is always smaller
+		 * than default_load_value. In this case the time elapsed until
+		 * than the timer restart was not yet added to
```
Comment (erwango) [inline drivers/timer/cortex_m_systick.c]: 'than' could be removed I guess

File: drivers/timer/cortex_m_systick.c [hunk 2]
```diff
@@ -753,7 +755,25 @@ return (u32_t) get_elapsed_count();
...
+		 * clock_accumulated_count. To retrieve a correct cycle count
+		 * we must therefore consider the number of cycle since current
+		 * tick period start and not only the cycle number since
+		 * the timer restart.
+		 */
+		if (SysTick->LOAD < default_load_value)
+			count = default_load_value;
```
Comment (erwango) [inline drivers/timer/cortex_m_systick.c]: Use {} even for one liner

File: drivers/timer/cortex_m_systick.c [hunk 3]
```diff
@@ -753,7 +755,25 @@ return (u32_t) get_elapsed_count();
...
+		else
+			count = SysTick->LOAD;
```
Comment (erwango) [inline drivers/timer/cortex_m_systick.c]: ditto

**Review by erwango — DISMISSED** [review]
Some minor comments

File: drivers/timer/cortex_m_systick.c (hunk 1 above)
Comment (greenhandholman) [reply drivers/timer/cortex_m_systick.c]: I removed it.

File: drivers/timer/cortex_m_systick.c (hunk 2 above)
Comment (greenhandholman) [reply drivers/timer/cortex_m_systick.c]: fix

File: drivers/timer/cortex_m_systick.c (hunk 3 above)
Comment (greenhandholman) [reply drivers/timer/cortex_m_systick.c]: fix

**Review by tyrantlizard77 — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
> Signed-off-by: Holman Greenhand greenhandholman@gmail.com

http://lotr.wikia.com/wiki/Holman_Greenhand

interesting :)

**Review by erwango — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**